onnxruntime/contrib_ops/cuda/collective/ipc_utils.cc

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "ipc_utils.h"
+#include "mpi_include.h"
+
+namespace onnxruntime {
+namespace cuda {
+namespace collective {
+
+#if defined(USE_MPI) || defined(USE_NCCL)
+
+using namespace onnxruntime;
+
+IpcMemory::IpcMemory(int rank, int world_size, std::size_t buffer_size)
+    : rank_(rank), world_size_(world_size), m_comm_ptrs_(world_size), mbuffer_size_(buffer_size) {
+  ORT_ENFORCE(AllocateIpcMemory() == Status::OK());
+}
+
+Status IpcMemory::AllocateIpcMemory() {
+  CUDA_RETURN_IF_ERROR(cudaMalloc(&m_buffer_ptr_, mbuffer_size_));
+  m_buffer_uptr_ = std::move(CudaMemPtrT{m_buffer_ptr_});
+
+  CUDA_RETURN_IF_ERROR(cudaMemset(m_buffer_ptr_, 0, mbuffer_size_));
+
+  cudaIpcMemHandle_t local_handle;
+  CUDA_RETURN_IF_ERROR(cudaIpcGetMemHandle(&local_handle, m_buffer_ptr_));
+
+  // Assume no pipeline parallelism.
+  InlinedVector<char> serial_handles(CUDA_IPC_HANDLE_SIZE * world_size_, 0);
+
+#ifdef USE_MPI
+  MPI_CHECK(MPI_Allgather(local_handle.reserved, CUDA_IPC_HANDLE_SIZE, MPI_BYTE, serial_handles.data(),
+                          CUDA_IPC_HANDLE_SIZE, MPI_BYTE, MPI_COMM_WORLD));
+#else
+  // TODO: Implement this for NCCL.
+  return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Please compile ORT with USE_MPI.");
+#endif
+
+  InlinedVector<cudaIpcMemHandle_t> handles(world_size_);
+  for (size_t i = 0; i < handles.size(); ++i) {
+    memcpy(handles[i].reserved, &serial_handles[i * CUDA_IPC_HANDLE_SIZE], CUDA_IPC_HANDLE_SIZE);
+  }
+
+  for (size_t node_id = 0; node_id < handles.size(); node_id++) {
+    if ((int)node_id == rank_) {
+      m_comm_ptrs_[node_id] = m_buffer_ptr_;
```

Comment:
>m_comm_ptrs_[node_id] = m_buffer_ptr_; [](http://example.com/codeflow?start=6&length=38)

I am still not clear about the purpose of storing different types of ptrs in the same array.

Reply:
In cuda kernels, it's natural to treat them as one array. (e.g peer_barrier_ptrs and peer_comm_buffer_ptrs), so we return one array.
But in IpcMemory class we split it into two arrays of smart pointer for memory management